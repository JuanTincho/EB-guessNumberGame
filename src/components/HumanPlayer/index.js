import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Button, Grid, Paper, TextField, Typography, withStyles,
} from '@material-ui/core';

export const messages = {
  win: 'You win!!!',
  smaller: 'My number is smaller',
  bigger: 'My number is bigger',
};

const styles = theme => ({
  button: {
    margin: theme.spacing.unit,
    textTransform: 'none',
  },
  paper: {
    padding: theme.spacing.unit * 2,
    textAlign: 'center',
    color: theme.palette.text.secondary,
  },
  textField: {
    marginLeft: theme.spacing.unit,
    marginRight: theme.spacing.unit,
  },
});

class HumanPlayer extends Component {
  state = {
    numberValue: 0,
    numberAnswer: Math.floor(Math.random() * 101),
    winState: false,
    message: '',
  };

  handleChangeNumber = (e) => {
    this.setState({
      numberValue: e.target.value,
    });
  };

  handleKeyPress = (e) => {
    if (e.which === 13) {
      this.checkAnswer();
    }
  };

  checkAnswer = () => {
    const { numberAnswer, numberValue } = this.state;

    const intNumberValue = parseInt(numberValue, 10);

    if (intNumberValue === numberAnswer) {
      this.setState({ winState: true });
    }

    if (intNumberValue <= numberAnswer) {
      this.setState({ message: messages.bigger });
    }

    if (intNumberValue >= numberAnswer) {
      this.setState({ message: messages.smaller });
    }
  };

  validateNumber = () => {
    const { numberValue } = this.state;
    return numberValue < 0 || numberValue > 100;
  };

  restartGame = () => {
    this.setState({
      winState: false,
      numberAnswer: Math.floor(Math.random() * 101),
      numberValue: 0,
      message: '',
    });
  };

  render() {
    const { classes } = this.props;
    const { message, numberValue, winState } = this.state;
    return (
      <Grid container justify="center">
        <Grid item xs={3}>
          <Paper className={classes.paper}>
            <>
              Choose a number between 1 and 100 <br /> and press Enter or click on the button below
            </>
          </Paper>
        </Grid>
        <Grid container justify="center">
          <TextField
            id="outlined"
            className={classes.textField}
            margin="normal"
            variant="outlined"
            type="number"
            label={this.validateNumber() ? 'Error' : 'Number'}
            error={this.validateNumber()}
            value={numberValue}
            onChange={this.handleChangeNumber}
            onKeyPress={this.handleKeyPress}
            inputProps={{ min: '0', max: '100', step: '1' }}
            disabled={winState}
          />
        </Grid>
        <Grid container justify="center">
          {!winState ? (
            <Button
              variant="contained"
              color="primary"
              className={classes.button}
              onClick={this.checkAnswer}
              disabled={this.validateNumber()}
            >
              Check Answer
            </Button>
          ) : (
            <Button
              variant="contained"
              color="secondary"
              className={classes.button}
              onClick={this.restartGame}
            >
              Play again
            </Button>
          )}
        </Grid>

        <Typography variant="h6" className={classes.paper} style={{ textAlign: 'center' }}>
          {!winState ? message : messages.win}
        </Typography>
      </Grid>
    );
  }
}

HumanPlayer.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(HumanPlayer);
