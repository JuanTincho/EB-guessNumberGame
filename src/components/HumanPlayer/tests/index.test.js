import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import HumanPlayer, { messages } from '../index';

describe('<HumanPlayer /> tests', () => {
  let component;
  let componentInstance;
  beforeEach(() => {
    component = shallow(<HumanPlayer />).dive();
    componentInstance = component.instance();
  });

  it('HumanPlayer shoud match snapshot', () => {
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Error label on TextField', () => {
    component.find('TextField').simulate('change', { target: { value: 105 } });

    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('handleKeyPress with EnterKey', () => {
    const mockedKeyValue = { which: 13 };
    componentInstance.checkAnswer = jest.fn();
    componentInstance.handleKeyPress(mockedKeyValue);
    expect(componentInstance.checkAnswer).toHaveBeenCalledTimes(1);
  });

  it('handleKeyPress with other KeyPress', () => {
    const mockedKeyValue = { e: { which: 14 } };
    componentInstance.checkAnswer = jest.fn();
    componentInstance.handleKeyPress(mockedKeyValue);
    expect(componentInstance.checkAnswer).toHaveBeenCalledTimes(0);
  });

  describe('checkAnswer', () => {
    it('should win the game when numberValue equals numberAnswer', () => {
      componentInstance.setState({ numberAnswer: 20, numberValue: '20' });
      expect(componentInstance.state.winState).toBeFalsy();
      componentInstance.checkAnswer();
      expect(componentInstance.state.winState).toBeTruthy();
    });

    it('should set the bigger message when numberValue is smaller than numberAnswer', () => {
      componentInstance.setState({ numberAnswer: 20, numberValue: '15' });
      expect(componentInstance.state.message).toEqual('');
      componentInstance.checkAnswer();
      expect(componentInstance.state.message).toEqual(messages.bigger);
    });

    it('should set the smaller message when numberValue is bigger than numberAnswer', () => {
      componentInstance.setState({ numberAnswer: 20, numberValue: '35' });
      expect(componentInstance.state.message).toEqual('');
      componentInstance.checkAnswer();
      expect(componentInstance.state.message).toEqual(messages.smaller);
    });
  });

  it('restartGame', () => {
    const mockedRandomNumber = 50;
    componentInstance.setState = jest.fn();
    Math.floor = jest.fn(() => mockedRandomNumber);
    const mockedValues = {
      winState: false,
      numberAnswer: mockedRandomNumber,
      numberValue: 0,
      message: '',
    };
    componentInstance.restartGame();
    expect(componentInstance.setState).toBeCalledTimes(1);
    expect(componentInstance.setState).toHaveBeenCalledWith(mockedValues);
  });
});
