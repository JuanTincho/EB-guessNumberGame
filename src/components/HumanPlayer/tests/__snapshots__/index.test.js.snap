// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HumanPlayer /> tests Error label on TextField 1`] = `
<div
  className="MuiGrid-container-4 MuiGrid-justify-xs-center-21"
>
  <div
    className="MuiGrid-item-5 MuiGrid-grid-xs-3-35"
  >
    <div
      className="MuiPaper-root-101 MuiPaper-elevation2-105 MuiPaper-rounded-102 HumanPlayer-paper-2"
    >
      Choose a number between 1 and 100
    </div>
  </div>
  <div
    className="MuiGrid-container-4 MuiGrid-justify-xs-center-21"
  >
    <div
      className="MuiFormControl-root-128 MuiFormControl-marginNormal-129 HumanPlayer-textField-3"
      onKeyPress={[Function]}
    >
      <label
        className="MuiFormLabel-root-143 MuiFormLabel-error-146 MuiInputLabel-error-135 MuiFormLabel-filled-147 MuiInputLabel-root-132 MuiInputLabel-formControl-137 MuiInputLabel-animated-140 MuiInputLabel-shrink-139 MuiInputLabel-outlined-142"
        data-shrink={true}
        htmlFor="outlined"
      >
        Error
      </label>
      <div
        className="MuiInputBase-root-163 MuiOutlinedInput-root-150 MuiInputBase-error-169 MuiOutlinedInput-error-155 MuiInputBase-formControl-164"
        onClick={[Function]}
      >
        <fieldset
          aria-hidden={true}
          className="MuiPrivateNotchedOutline-root-180 MuiOutlinedInput-notchedOutline-157"
          style={
            Object {
              "paddingLeft": 8,
            }
          }
        >
          <legend
            className="MuiPrivateNotchedOutline-legend-181"
            style={
              Object {
                "width": 0,
              }
            }
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
        <input
          aria-invalid={true}
          className="MuiInputBase-input-173 MuiOutlinedInput-input-158 MuiInputBase-inputType-176"
          disabled={false}
          id="outlined"
          max="100"
          min="0"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          step="1"
          type="number"
          value={105}
        />
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-container-4 MuiGrid-justify-xs-center-21"
  >
    <button
      className="MuiButtonBase-root-208 MuiButtonBase-disabled-209 MuiButton-root-182 MuiButton-contained-193 MuiButton-containedPrimary-194 MuiButton-raised-196 MuiButton-raisedPrimary-197 MuiButton-disabled-202 HumanPlayer-button-1"
      disabled={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="-1"
      type="button"
    >
      <span
        className="MuiButton-label-183"
      >
        Check Answer
      </span>
    </button>
  </div>
  <h6
    className="MuiTypography-root-211 MuiTypography-h6-228 HumanPlayer-paper-2"
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    
  </h6>
</div>
`;

exports[`<HumanPlayer /> tests HumanPlayer shoud match snapshot 1`] = `
<div
  className="MuiGrid-container-4 MuiGrid-justify-xs-center-21"
>
  <div
    className="MuiGrid-item-5 MuiGrid-grid-xs-3-35"
  >
    <div
      className="MuiPaper-root-101 MuiPaper-elevation2-105 MuiPaper-rounded-102 HumanPlayer-paper-2"
    >
      Choose a number between 1 and 100
    </div>
  </div>
  <div
    className="MuiGrid-container-4 MuiGrid-justify-xs-center-21"
  >
    <div
      className="MuiFormControl-root-128 MuiFormControl-marginNormal-129 HumanPlayer-textField-3"
      onKeyPress={[Function]}
    >
      <label
        className="MuiFormLabel-root-143 MuiFormLabel-filled-147 MuiInputLabel-root-132 MuiInputLabel-formControl-137 MuiInputLabel-animated-140 MuiInputLabel-shrink-139 MuiInputLabel-outlined-142"
        data-shrink={true}
        htmlFor="outlined"
      >
        Number
      </label>
      <div
        className="MuiInputBase-root-163 MuiOutlinedInput-root-150 MuiInputBase-formControl-164"
        onClick={[Function]}
      >
        <fieldset
          aria-hidden={true}
          className="MuiPrivateNotchedOutline-root-180 MuiOutlinedInput-notchedOutline-157"
          style={
            Object {
              "paddingLeft": 8,
            }
          }
        >
          <legend
            className="MuiPrivateNotchedOutline-legend-181"
            style={
              Object {
                "width": 0,
              }
            }
          >
            <span
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
        <input
          aria-invalid={false}
          className="MuiInputBase-input-173 MuiOutlinedInput-input-158 MuiInputBase-inputType-176"
          disabled={false}
          id="outlined"
          max="100"
          min="0"
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          step="1"
          type="number"
          value={0}
        />
      </div>
    </div>
  </div>
  <div
    className="MuiGrid-container-4 MuiGrid-justify-xs-center-21"
  >
    <button
      className="MuiButtonBase-root-208 MuiButton-root-182 MuiButton-contained-193 MuiButton-containedPrimary-194 MuiButton-raised-196 MuiButton-raisedPrimary-197 HumanPlayer-button-1"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex="0"
      type="button"
    >
      <span
        className="MuiButton-label-183"
      >
        Check Answer
      </span>
      <span
        className="MuiTouchRipple-root-246"
      />
    </button>
  </div>
  <h6
    className="MuiTypography-root-211 MuiTypography-h6-228 HumanPlayer-paper-2"
    style={
      Object {
        "textAlign": "center",
      }
    }
  >
    
  </h6>
</div>
`;
